torch/csrc/jit/script/module.h

```diff
@@ -125,17 +126,117 @@ struct TORCH_API Method {
   // first-class function in class_compilation_unit()
   Function* function_;
 };
 
-struct TORCH_API Module {
+struct TORCH_API Object {
+  Object() {}
+  Object(ObjectPtr object_value) : object_value_(std::move(object_value)) {}
+
+  const c10::QualifiedName& name() const {
+    return *object_value()->type()->name();
+  }
+
+  void setattr(const std::string& name, IValue v) {
+    size_t slot = object_value()->type()->getAttributeSlot(name);
+    const TypePtr& expected = object_value()->type()->getAttribute(slot);
+    TORCH_CHECK(
+        v.type()->isSubtypeOf(expected),
+        "Expected a value of type '",
+        expected->python_str(),
+        "' for field '",
+        name,
+        "', but found '",
+        v.type()->python_str(),
+        "'");
+    object_value()->setSlot(slot, std::move(v));
+  }
+
+  ObjectPtr object_value() const;
+
+  IValue attr(const std::string& name) const {
+    return object_value()->getAttr(name);
+  }
+
+  IValue attr(const std::string& name, IValue or_else) const {
+    if (auto r = object_value()->type()->findAttributeSlot(name)) {
+      return object_value()->getSlot(*r);
+    }
+    return or_else;
+  }
+
+  bool hasattr(const std::string& name) const {
+    return object_value()->type()->findAttributeSlot(name).has_value();
+  }
+
+  // each module owns its method. The reference returned here
+  // is guarenteed to stay valid until this module has been destroyed
+  Method get_method(const std::string& name) const {
+    if (auto method = find_method(name)) {
+      return *method;
+    }
+    AT_ERROR("Method '", name, "' is not defined.");
+  }
+
+  c10::optional<Method> find_method(const std::string& basename) const;
+
+  const std::vector<Method> get_methods() const {
+    return fmap(type()->methods(), [&](Function* func) {
+      return Method(object_value(), func);
+    });
+  }
+
+  /// Run a method from this module.
+  ///
+  /// For example:
+  /// @code
+  ///   IValue output = module->run("relu_script", a, b);
+  /// @endcode
+  ///
+  /// To get a compile a module from a source string, see torch::jit::compile
+  ///
+  /// @param method_name The name of the method to run
+  /// @param args Arguments to be passed to the method
+  /// @return An IValue containing the return value (or values if it is a tuple)
+  /// from the method
+  template <typename... Types>
+  IValue run_method(const std::string& method_name, Types&&... args) {
+    return get_method(method_name)({IValue(std::forward<Types>(args))...});
+  }
+
+  ClassTypePtr type() const {
+    return object_value()->type();
+  }
+  std::shared_ptr<CompilationUnit> class_compilation_unit() const {
+    return object_value()->compilation_unit();
+  }
+
+  // so that C++ users can easily add methods
+  void define(const std::string& src, const ResolverPtr& resolver = nullptr);
+
+  template <typename... Types>
+  IValue create_class(const c10::QualifiedName& name, Types&&... args) const {
```

Comment:
We shouldn't bind `create_class` into Object. First it probably should be called `create_object` because it is not creating a new type. Second it is really only creating an object in the same compilation unit as this object. I am not sure that should be a method on Object. We can keep these calls on Module until we have a better place for them.